text/0000-single-file-components.md

```diff
@@ -0,0 +1,111 @@
+- Start Date: 2017-05-22
+- RFC PR: (leave this empty)
+- Ember Issue: (leave this empty)
+
+# Summary
+
+It is currently possible to create single file Ember components with component
+templates declared inline using a tagged template string assinged to the `layout`
+property. However, this pattern is not documented and support is unknown. This RFC
+proposes documenting support for this pattern.
+
+# Motivation
+
+The benefits of same file template declarations are mostly development conveniences
+focused around:
+- All component concerns available in a single file, making tracking property or
+  service usage throughout a component easier.
+- Easier component file portability when restructuring components/creating namespaces
+  for components.
+
+Anecdotally our team has also found that single file components support focused
+modular design, as it's easy to 'feel' when your component is trying to address too
+many concerns as the file length grows.
+
+Primarily though, single file components are a convenience that makes working with
+numerous components at once easier to manage. I think this is a benefit of Vue and
+React that is already available in Ember as well.
+
+# Detailed design
+
+This pattern is currently possible by assigning a tagged template string to the
+component layout property:
+
+```javascript
+import Component from 'ember-component';
+import inject from 'ember-service/inject';
+import hbs from 'htmlbars-inline-precompile';
+
+export default Component.extend({
+  cart: inject()
+
+  passedItemPrice: 0,
+
+  layout: hbs`
+    <div>
+      <span>Item cost is: {{total}}</span>
+      <button {{action 'removeFromCart' target=cart}}>Remove</button>
+    </div>
+  `
+})
+```
+
+For proper syntax highlighting of handlebars inside the tagged template literal our
+team has created highlighters for
+[Sublime](https://github.com/healthsparq/sublime-ember-syntax),
+[Atom](https://atom.io/packages/language-ember), and
+[VSCode](https://marketplace.visualstudio.com/items?itemName=dhedgecock.ember-syntax).
+
+A proposed change to this pattern is allowing the component template to be declared
```

Comment:
I strongly agree with the goal of this paragraph. I also agree with @locks that it should be done with care.

What is currently informally called "layout" and "template" should become "template" and "block". There's a few reasons for this.

1) I think that when most people think of a template they think of an entire `hbs` file.
2) When people say "component's layout", they also think of a template. I think it's fine to use both words when talking about components informally, but the docs should use "template".
2) Everyone already calls blocks "blocks".
3) Moreover, named blocks are... "blocks" :D